cmd/cli/operations/update_job.go

```diff
@@ -0,0 +1,173 @@
+		filePath := dir + "/" + f.Name()
+		fi, err := o.Filesystem.Stat(filePath)
+		if err != nil {
+			return "", err
+		}
+		currTime := fi.ModTime().Unix()
+		if currTime > newestTime {
+			newestTime = currTime
+			newestFile = filePath
+		}
+	}
+
+	return newestFile, nil
+}
+
+func (o RealOperator) filterRunningJobs(jobs []flink.Job) (ret []flink.Job) {
+	for _, job := range jobs {
+		if job.Status == "RUNNING" {
+			ret = append(ret, job)
+		}
+	}
+	return
+}
+
+func (o RealOperator) monitorSavepointCreation(jobID string, requestID string, maxElapsedTime int) error {
+	op := func() error {
+		log.Println("checking status of savepoint creation")
+		res, err := o.FlinkRestAPI.MonitorSavepointCreation(jobID, requestID)
+		if err != nil {
+			return err
+		}
+
+		switch res.Status.Id {
+		case "COMPLETED":
+			return nil
+		case "IN_PROGRESS":
+			return fmt.Errorf("savepoint creation for job \"%v\" is still pending", jobID)
+		default:
+			return fmt.Errorf("savepoint creation for job \"%v\" returned an unknown status \"%v\"", jobID, res.Status)
+		}
+	}
+	b := &backoff.ExponentialBackOff{
+		InitialInterval:     backoff.DefaultInitialInterval,
+		RandomizationFactor: backoff.DefaultRandomizationFactor,
+		Multiplier:          backoff.DefaultMultiplier,
+		MaxInterval:         backoff.DefaultMaxInterval,
+		MaxElapsedTime:      time.Duration(maxElapsedTime) * time.Second,
+		Clock:               backoff.SystemClock,
+	}
+	err := backoff.Retry(op, b)
+	if err != nil {
+		return fmt.Errorf("failed to create savepoint for job \"%v\" within %v seconds", jobID, b.MaxElapsedTime.Seconds())
+	}
+
+	b.Reset()
+
+	return nil
+}
+
+// Update executes the actual update of a job on the Flink cluster
+func (o RealOperator) Update(u UpdateJob) error {
+	if len(u.JobNameBase) == 0 {
+		return errors.New("unspecified argument 'JobNameBase'")
+	}
+	if len(u.SavepointDir) == 0 {
+		return errors.New("unspecified argument 'SavepointDir'")
+	}
+
+	log.Printf("starting job update for base name '%v' and savepoint dir '%v'\n", u.JobNameBase, u.SavepointDir)
+
+	jobs, err := o.FlinkRestAPI.RetrieveJobs()
+	if err != nil {
+		return fmt.Errorf("retrieving jobs failed: %v", err)
+	}
+
+	runningJobs := o.filterRunningJobs(jobs)
+
+	deploy := Deploy{
+		LocalFilename:         u.LocalFilename,
+		RemoteFilename:        u.RemoteFilename,
+		ApiToken:              u.APIToken,
+		EntryClass:            u.EntryClass,
+		Parallelism:           u.Parallelism,
+		ProgramArgs:           u.ProgramArgs,
+		AllowNonRestoredState: u.AllowNonRestoredState,
+	}
+	switch len(runningJobs) {
+	case 0:
+		log.Printf("no instance running for job name base \"%v\". Deploying with last available savepoint\n", u.JobNameBase)
+	case 1:
+		log.Printf("found exactly 1 running job with base name: %v\n", u.JobNameBase)
+		job := runningJobs[0]
+
+		log.Printf("creating savepoint for job \"%v\"", job.ID)
+		savepointResponse, err := o.FlinkRestAPI.CreateSavepoint(job.ID, u.SavepointDir)
```

Comment:
Would be nice if we can somehow get the savepoint location without specifying the savepoint dir here. Then we wouldn't need the savepointDir here at all.
Just a placeholder to maybe check one last time tomorrow if we can get it out of the API somewhere